### PR TITLE
Fix #14844

### LINKED ISSUE #14844: Illegal symbol reuse in template with method call syntax
This snippet fails:
```nim
template makeSeq: untyped =
  var i = 0
  for _ in 0..<10: # Already fails with 0..<6 which is exactly 10 / 2 + 1
    assert i in 0..<10 # This assertion fails
    inc i
  @[1] # Works with [1]

template last(s): untyped = s[s.len - 1] # Works with s[^1]

echo makeSeq().last() # This doesn't work
#echo last(makeSeq()) # This works
```

### Current Output
```
test.nim(4)          test
lib/system/assertions.nim(30) failedAssertImpl
lib/system/assertions.nim(23) raiseAssert
lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: test.nim(4, 12) `i`gensym4607016 in 0 ..< 10`  [AssertionDefect]
Error: execution of an external program failed: 'test '
```

### Expected Output
```
1
```

I haven't invesitgated yet (only minimized), but the fact that it fails when the for loop range is bigger than half of the assertion range leads me to believe that the for loop is running "too far".
One explanation that comes to mind is that the for loop is getting evaluated twice because of the `last` template, but when used with UFCS the second for loop somehow ends up using the first for loop's `i`, instead of its own.
EDIT: Confirmed by replacing the `assert` with `echo i`.

### Additional information:
```
$ nim -v
Nim Compiler Version #devel
https://github.com/nim-lang/Nim/commit/62394616e848e7a94bc0208d6535df6f582e74ce
```
This bug is the actual rootcause of #14404

### PR DESCRIPTION
**Pull request by Clyybber**
One test will fail because the current broken behaviour allowed it to (https://play.nim-lang.org/#ix=2qAU).
Should both fail or work? A manual `(var i = 0); (var i = 0)` currently fails.. (for more context see https://github.com/nim-lang/Nim/commit/bb532a697edad1bac60a87a7ff43956c9635973d#diff-705c21e2cbb7aca092a9882547994008)

### CODE REVIEW

**Comment by timotheecour** [comment]
* could you add tests cases, at least the ones in https://github.com/nim-lang/Nim/issues/14844#issuecomment-651295359 ? (ideally turning them into doAssert tests)

* likewise with top-level vs proc scope tests mentioned there

> Should both fail or work? 

given that all these fail:
```nim
when defined case6:
  template bar*() =
    var r = @[(var i = @[""]; i), (var i = @[""]; i)]
    echo r
  bar()

when defined case7:
  var r = @[(var i = @[""]; i), (var i = @[""]; i)]
  echo r

when defined case8:
  proc main=
    var r = @[(var i = @[""]; i), (var i = @[""]; i)]
    echo r
  main()
```

it's better to be consistent; as well as to be consistent between MCS and regular call syntax, so the logical thing to do is make both fail; if absolutely needed add a `-d:nimHasWorkaround14863` for a transition time. 

Note that I'm fixing `applyIt` here to avoid the double evaluation bug: https://github.com/nim-lang/Nim/pull/14869

**Comment by timotheecour** [comment]
can `nkEarlySemArg` also be used to solve the long standing issue of const folding happening too early, eg:
* https://github.com/nim-lang/Nim/issues/9301
* https://github.com/nim-lang/Nim/issues/6293
* https://github.com/nim-lang/Nim/issues/11789
?

File: compiler/ast.nim [hunk 1]
```diff
@@ -225,6 +225,7 @@ type
     nkBreakState,         # special break statement for easier code generation
     nkFuncDef,            # a func
     nkTupleConstr         # a tuple constructor
+    nkEarlySemArg         # When an argument was semmed early, but we might need the untyped AST later, the semmed AST is the first, the unsemmed the second son
```
Comment (Araq) [inline compiler/ast.nim:228]: Simply remembering both variants can create a compiler that uses an expontential amount of memory. (Nest these calls to get 2x2x2x2x... memory consumption.) There should only be a single AST that works with all the constructs. My "semtityped" AST idea.

**Comment by Araq** [comment]
This is very dangerous terrain.

File: compiler/ast.nim (hunk 1 above)
Comment (Clyybber) [reply compiler/ast.nim:228]: No worries, I have an unpushed fix for that.

Comment (Araq) [reply compiler/ast.nim:228]: *Now* I worry...

Comment (Clyybber) [reply compiler/ast.nim:228]: We just discard our reference to the unsemmed/semmed AST when choosing which one we need after sigmatch. https://github.com/nim-lang/Nim/pull/14863/commits/fb0e697bbdf20a4087dfd463ab018d2fbb1f683e

**Comment by Araq** [comment]
Too convoluted. Instead sem'checking should understand nkHiddenAddr, nkHiddenDeref and we need a `nkHiddenArrayConstr` node to make sem'checking idempotent.

**Comment by Clyybber** [comment]
That isn't related. This is about preventing typed AST ending up in untyped arguments.

**Comment by Araq** [comment]
And why is that a problem? "untyped" must go away, we need a solid typed AST so that macros can do type-based analysis. We don't need "I need untyped because the AST is then simpler", that's a symptom of the underlying issues.

**Pull request closed without merging**